From linux-kernel-owner+chrisw=40osdl.org-S261317AbVFAHlL@vger.kernel.org  Wed Jun  1 00:42:56 2005
Date: 	Wed, 1 Jun 2005 09:40:59 +0200
From: Jan Kara <jack@suse.cz>
To: akpm@osdl.org
Cc: sct@redhat.com, linux-kernel@vger.kernel.org
Subject: [PATCH] ext3: fix log_do_checkpoint() assertion failure

Fix possible false assertion failure in log_do_checkpoint(). We might fail
to detect that we actually made a progress when cleaning up the checkpoint
lists if we don't retry after writing something to disk. The patch was
confirmed to fix observed assertion failures for several users.

When we flushed some buffers we need to retry scanning the list.
Otherwise we can fail to detect our progress.

Signed-off-by: Jan Kara <jack@suse.cz>
Signed-off-by: Andrew Morton <akpm@osdl.org>
Signed-off-by: Chris Wright <chrisw@osdl.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 fs/jbd/checkpoint.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletion(-)

--- linux-2.6.11.10.orig/fs/jbd/checkpoint.c	2005-05-16 10:50:47.000000000 -0700
+++ linux-2.6.11.10/fs/jbd/checkpoint.c	2005-06-08 22:40:33.000000000 -0700
@@ -339,8 +339,10 @@
 			}
 		} while (jh != last_jh && !retry);
 
-		if (batch_count)
+		if (batch_count) {
 			__flush_batch(journal, bhs, &batch_count);
+			retry = 1;
+		}
 
 		/*
 		 * If someone cleaned up this transaction while we slept, we're
